kinesis/kinesis.go

```diff
@@ -0,0 +1,311 @@
+        }
+        svcConfig.EndpointResolver = endpoints.ResolverFunc(cwCustomResolverFn)
+    }
+
+    if roleARN != "" {
+        creds := stscreds.NewCredentials(sess, roleARN)
+        svcConfig.Credentials = creds
+    }
+
+    client := kinesis.New(sess, svcConfig)
+    client.Handlers.Build.PushBackNamed(plugins.CustomUserAgentHandler())
+    return client
+}
+
+// AddRecord accepts a record and adds it to the buffer, flushing the buffer if it is full
+// the return value is one of: FLB_OK FLB_RETRY
+// API Errors lead to an FLB_RETRY, and all other errors are logged, the record is discarded and FLB_OK is returned
+func (outputPlugin *OutputPlugin) AddRecord(record map[interface{}]interface{}) int {
+    partitionKey := outputPlugin.getPartitionKey(record)
+
+    data, err := outputPlugin.processRecord(record)
+    if err != nil {
+        logrus.Errorf("[kinesis %d] %v\n", outputPlugin.PluginID, err)
+        // discard this single bad record instead and let the batch continue
+        return fluentbit.FLB_OK
+    }
+
+    newDataSize := len(data)
+
+    if len(outputPlugin.records) == maximumRecordsPerPut || (outputPlugin.dataLength+newDataSize) > maximumPutRecordBatchSize {
+        err = outputPlugin.sendCurrentBatch()
+        if err != nil {
+            logrus.Errorf("[kinesis %d] %v\n", outputPlugin.PluginID, err)
+            // send failures are retryable
+            return fluentbit.FLB_RETRY
+        }
+    }
+
+    outputPlugin.records = append(outputPlugin.records, &kinesis.PutRecordsRequestEntry{
+        Data: data,
+        PartitionKey: aws.String(partitionKey),
+    })
+    outputPlugin.dataLength += newDataSize
+    return fluentbit.FLB_OK
+}
+
+// Flush sends the current buffer of records
+func (outputPlugin *OutputPlugin) Flush() error {
+    return outputPlugin.sendCurrentBatch()
+}
+
+func (outputPlugin *OutputPlugin) processRecord(record map[interface{}]interface{}) ([]byte, error) {
+    if outputPlugin.dataKeys != "" {
+        record = plugins.DataKeys(outputPlugin.dataKeys, record)
+    }
+
+    var err error
+    record, err = plugins.DecodeMap(record)
+    if err != nil {
+        logrus.Debugf("[kinesis %d] Failed to decode record: %v\n", outputPlugin.PluginID, record)
+        return nil, err
+    }
+
+    var json = jsoniter.ConfigCompatibleWithStandardLibrary
+    data, err := json.Marshal(record)
+    if err != nil {
+        logrus.Debugf("[kinesis %d] Failed to marshal record: %v\n", outputPlugin.PluginID, record)
+        return nil, err
+    }
+
+    // append newline
+    data = append(data, []byte("\n")...)
+
+    if len(data) > maximumRecordSize {
+        return nil, fmt.Errorf("Log record greater than max size allowed by Kinesis")
+    }
+
+    return data, nil
+}
+
+func (outputPlugin *OutputPlugin) sendCurrentBatch() error {
+    if outputPlugin.partitionKeyError != nil {
+        logData := outputPlugin.records[0].Data
```

Comment:
This will work 99% of the time, but I'd prefer if you got the log record that failed when you call the getPartitionKey() function-

Reply:
Saving the index for which data record we are getting the invalid partition key, and printig the exact log record.